File: ios/Runner/AmapGlobalPlatformView.swift
#if canImport(Flutter)
import Flutter
#endif
import Foundation
#if canImport(MAMapKit)
import MAMapKit
#endif
#if canImport(UIKit)
import UIKit
#endif
#if canImport(CoreLocation)
import CoreLocation
#endif

#if canImport(Flutter) && canImport(UIKit)

/// Registers a UIKit backed platform view for displaying global cities on AMap.
final class AmapGlobalViewFactory: NSObject, FlutterPlatformViewFactory {
  private let messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    FlutterStandardMessageCodec.sharedInstance()
  }

  func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
    AmapGlobalPlatformView(frame: frame, viewId: viewId, arguments: args, messenger: messenger)
  }
}

/// City marker data model
struct CityMarkerData {
  let id: Int
  let name: String
  let latitude: Double
  let longitude: Double
  let country: String
  let score: Double
}

/// Platform view displaying global city markers on AMap.
final class AmapGlobalPlatformView: NSObject, FlutterPlatformView {
  private let containerView: UIView
  private let mapView: MAMapView
  private let methodChannel: FlutterMethodChannel
  private var cities: [CityMarkerData] = []
  private var annotations: [MAPointAnnotation] = []

  init(frame: CGRect, viewId: Int64, arguments args: Any?, messenger: FlutterBinaryMessenger) {
    // 创建容器视图
    containerView = UIView(frame: frame)
    containerView.backgroundColor = UIColor.white
    containerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    
    // 创建地图视图 - 确保有足够的初始 frame
    let mapFrame = CGRect(x: 0, y: 0, width: max(frame.width, 300), height: max(frame.height, 500))
    mapView = MAMapView(frame: mapFrame)
    mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    mapView.showsUserLocation = false
    mapView.showsCompass = true
    mapView.showsScale = true
    
    // 使用标准地图类型（2D），在模拟器上更兼容
    mapView.mapType = MAMapType.standard
    
    // 设置方法通道用于 Flutter 和原生通信
    methodChannel = FlutterMethodChannel(
      name: "amap_global_view_\(viewId)",
      binaryMessenger: messenger
    )

    super.init()

    // 将地图添加到容器
    containerView.addSubview(mapView)
    
    mapView.delegate = self
    
    // 设置方法通道处理
    setupMethodChannel()
    
    // 配置地图和城市标记
    configureMap(arguments: args)
    
    NSLog("🗺️ AmapGlobalPlatformView initialized with frame: \(frame), mapFrame: \(mapFrame)")
  }

  func view() -> UIView {
    containerView
  }

  private func setupMethodChannel() {
    methodChannel.setMethodCallHandler { [weak self] (call, result) in
      guard let self = self else { return }
      
      switch call.method {
      case "setZoom":
        if let args = call.arguments as? [String: Any],
           let zoom = args["zoom"] as? Double {
          self.mapView.setZoomLevel(CGFloat(zoom), animated: true)
          result(nil)
        } else {
          result(FlutterError(code: "INVALID_ARGS", message: "Zoom level required", details: nil))
        }
        
      case "setCenter":
        if let args = call.arguments as? [String: Any],
           let lat = args["latitude"] as? Double,
           let lng = args["longitude"] as? Double {
          let coordinate = CLLocationCoordinate2D(latitude: lat, longitude: lng)
          self.mapView.setCenter(coordinate, animated: true)
          result(nil)
        } else {
          result(FlutterError(code: "INVALID_ARGS", message: "Latitude and longitude required", details: nil))
        }
        
      case "resetToWorld":
        // 重置到世界视图
        let worldCenter = CLLocationCoordinate2D(latitude: 20.0, longitude: 0.0)
        self.mapView.setCenter(worldCenter, animated: true)
        self.mapView.setZoomLevel(2, animated: true)
        result(nil)
        
      case "updateCities":
        if let args = call.arguments as? [String: Any],
           let citiesData = args["cities"] as? [[String: Any]] {
          self.updateCities(citiesData: citiesData)
          result(nil)
        } else {
          result(FlutterError(code: "INVALID_ARGS", message: "Cities data required", details: nil))
        }
        
      default:
        result(FlutterMethodNotImplemented)
      }
    }
  }

  private func configureMap(arguments args: Any?) {
    guard let params = args as? [String: Any] else {
      NSLog("🗺️ No arguments provided, using defaults")
      setDefaultMapPosition()
      return
    }

    NSLog("🗺️ Configuring map with params: \(params)")

    // 设置初始中心点 - 默认使用中国中心以确保底图可见
    let centerLat = params["centerLatitude"] as? Double ?? 35.0
    let centerLng = params["centerLongitude"] as? Double ?? 105.0
    let centerCoordinate = CLLocationCoordinate2D(latitude: centerLat, longitude: centerLng)
    mapView.setCenter(centerCoordinate, animated: false)
    NSLog("🗺️ Set center: \(centerLat), \(centerLng)")

    // 设置初始缩放 - 使用较高的缩放级别确保底图可见
    if let initialZoom = params["initialZoom"] as? Double {
      // 确保缩放级别不会太小
      let safeZoom = max(initialZoom, 3.0)
      mapView.setZoomLevel(CGFloat(safeZoom), animated: false)
      NSLog("🗺️ Set zoom level: \(safeZoom)")
    } else {
      mapView.setZoomLevel(4, animated: false)
    }

    // 添加城市标记
    if let citiesData = params["cities"] as? [[String: Any]] {
      NSLog("🗺️ Received \(citiesData.count) cities")
      updateCities(citiesData: citiesData)
    }
  }

  private func setDefaultMapPosition() {
    // 默认显示中国视图 - 高德地图在中国区域底图覆盖完整
    let chinaCenter = CLLocationCoordinate2D(latitude: 35.0, longitude: 105.0)
    mapView.setCenter(chinaCenter, animated: false)
    mapView.setZoomLevel(4, animated: false)  // 使用更高的缩放级别确保底图可见
    NSLog("🗺️ Set default map position to China center")
  }

  private func updateCities(citiesData: [[String: Any]]) {
    // 清除现有标记
    mapView.removeAnnotations(annotations)
    annotations.removeAll()
    cities.removeAll()

    // 解析城市数据
    for cityData in citiesData {
      guard let id = cityData["id"] as? Int,
            let name = cityData["name"] as? String,
            let lat = cityData["latitude"] as? Double,
            let lng = cityData["longitude"] as? Double else {
        continue
      }

      let city = CityMarkerData(
        id: id,
        name: name,
        latitude: lat,
        longitude: lng,
        country: cityData["country"] as? String ?? "",
        score: cityData["score"] as? Double ?? 0.0
      )
      cities.append(city)

      // 创建标记
      let annotation = MAPointAnnotation()
      annotation.coordinate = CLLocationCoordinate2D(latitude: lat, longitude: lng)
      annotation.title = name
      annotation.subtitle = city.country
      annotations.append(annotation)
    }

    // 添加标记到地图
    mapView.addAnnotations(annotations)

    NSLog("🗺️ Added \(annotations.count) city markers to AMap")
  }
}

// MARK: - MAMapViewDelegate

extension AmapGlobalPlatformView: MAMapViewDelegate {
  
  func mapViewDidFinishLoadingMap(_ mapView: MAMapView!) {
    NSLog("✅ AMap finished loading map tiles")
  }
  
  func mapViewDidFailLoadingMap(_ mapView: MAMapView!, withError error: Error!) {
    NSLog("❌ AMap failed to load map: \(error?.localizedDescription ?? "unknown error")")
  }
  
  func mapInitComplete(_ mapView: MAMapView!) {
    NSLog("✅ AMap init complete")
  }
  
  func mapView(_ mapView: MAMapView!, viewFor annotation: MAAnnotation!) -> MAAnnotationView! {
    guard !(annotation is MAUserLocation) else { return nil }

    let identifier = "CityMarker"
    var annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: identifier) as? MAPinAnnotationView

    if annotationView == nil {
      annotationView = MAPinAnnotationView(annotation: annotation, reuseIdentifier: identifier)
      annotationView?.canShowCallout = true
      annotationView?.animatesDrop = false
      annotationView?.pinColor = .red
    } else {
      annotationView?.annotation = annotation
    }

    // 根据分数设置不同颜色
    if let pointAnnotation = annotation as? MAPointAnnotation,
       let index = annotations.firstIndex(of: pointAnnotation),
       index < cities.count {
      let city = cities[index]
      if city.score >= 4.0 {
        annotationView?.pinColor = .green
      } else if city.score >= 3.0 {
        annotationView?.pinColor = .purple
      } else {
        annotationView?.pinColor = .red
      }
    }

    return annotationView
  }

  func mapView(_ mapView: MAMapView!, didSelect view: MAAnnotationView!) {
    guard let annotation = view.annotation,
          !(annotation is MAUserLocation) else { return }

    // 发送城市点击事件到 Flutter
    if let pointAnnotation = annotation as? MAPointAnnotation,
       let index = annotations.firstIndex(of: pointAnnotation),
       index < cities.count {
      let city = cities[index]
      methodChannel.invokeMethod("onCityTapped", arguments: [
        "id": city.id,
        "name": city.name,
        "latitude": city.latitude,
        "longitude": city.longitude,
        "country": city.country,
        "score": city.score
      ])
    }
  }

  func mapView(_ mapView: MAMapView!, mapDidZoomByUser wasUserAction: Bool) {
    // 可以通知 Flutter 缩放级别变化
    if wasUserAction {
      methodChannel.invokeMethod("onZoomChanged", arguments: [
        "zoom": mapView.zoomLevel
      ])
    }
  }
}

#endif

File: android/app/src/main/kotlin/com/example/df_admin_mobile/AmapGlobalPlatformView.kt
package com.example.df_admin_mobile

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Color
import android.view.View
import com.amap.api.maps.AMap
import com.amap.api.maps.CameraUpdateFactory
import com.amap.api.maps.MapView
import com.amap.api.maps.model.BitmapDescriptorFactory
import com.amap.api.maps.model.CameraPosition
import com.amap.api.maps.model.LatLng
import com.amap.api.maps.model.Marker
import com.amap.api.maps.model.MarkerOptions
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

/**
 * 高德地图全球视图工厂
 */
class AmapGlobalViewFactory(private val messenger: BinaryMessenger) : 
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val creationParams = args as? Map<String, Any>
        return AmapGlobalPlatformView(context, viewId, creationParams, messenger)
    }
}

/**
 * 城市标记数据类
 */
data class CityMarkerData(
    val id: Int,
    val name: String,
    val latitude: Double,
    val longitude: Double,
    val country: String,
    val score: Double
)

/**
 * 高德地图全球 Platform View - 显示全球城市分布
 */
class AmapGlobalPlatformView(
    private val context: Context,
    private val viewId: Int,
    private val creationParams: Map<String, Any>?,
    messenger: BinaryMessenger
) : PlatformView, MethodChannel.MethodCallHandler {
    
    private val mapView: MapView = MapView(context)
    private var aMap: AMap? = null
    private val methodChannel: MethodChannel
    private val cities = mutableListOf<CityMarkerData>()
    private val markers = mutableListOf<Marker>()
    
    init {
        // 初始化方法通道
        methodChannel = MethodChannel(messenger, "amap_global_view_$viewId")
        methodChannel.setMethodCallHandler(this)
        
        // 初始化地图
        mapView.onCreate(null)
        aMap = mapView.map
        
        // 配置地图
        configureMap()
    }
    
    private fun configureMap() {
        aMap?.apply {
            // 地图基本设置
            uiSettings.isZoomControlsEnabled = false
            uiSettings.isCompassEnabled = true
            uiSettings.isScaleControlsEnabled = true
            uiSettings.isMyLocationButtonEnabled = false
            
            // 设置地图类型
            mapType = AMap.MAP_TYPE_NORMAL
            
            // 设置初始位置和缩放
            val initialZoom = (creationParams?.get("initialZoom") as? Double)?.toFloat() ?: 2f
            val centerLat = creationParams?.get("centerLatitude") as? Double ?: 20.0
            val centerLng = creationParams?.get("centerLongitude") as? Double ?: 0.0
            
            val cameraPosition = CameraPosition.Builder()
                .target(LatLng(centerLat, centerLng))
                .zoom(initialZoom)
                .build()
            moveCamera(CameraUpdateFactory.newCameraPosition(cameraPosition))
            
            // 设置标记点击监听
            setOnMarkerClickListener { marker ->
                onMarkerClicked(marker)
                true
            }
            
            // 解析并添加城市标记
            @Suppress("UNCHECKED_CAST")
            val citiesData = creationParams?.get("cities") as? List<Map<String, Any>>
            citiesData?.let { updateCities(it) }
        }
    }
    
    override fun getView(): View = mapView
    
    override fun dispose() {
        mapView.onDestroy()
    }
    
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setZoom" -> {
                val zoom = call.argument<Double>("zoom")?.toFloat()
                if (zoom != null) {
                    aMap?.animateCamera(CameraUpdateFactory.zoomTo(zoom))
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Zoom level required", null)
                }
            }
            "setCenter" -> {
                val lat = call.argument<Double>("latitude")
                val lng = call.argument<Double>("longitude")
                if (lat != null && lng != null) {
                    aMap?.animateCamera(CameraUpdateFactory.newLatLng(LatLng(lat, lng)))
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Latitude and longitude required", null)
                }
            }
            "resetToWorld" -> {
                val worldPosition = CameraPosition.Builder()
                    .target(LatLng(20.0, 0.0))
                    .zoom(2f)
                    .build()
                aMap?.animateCamera(CameraUpdateFactory.newCameraPosition(worldPosition))
                result.success(null)
            }
            "updateCities" -> {
                @Suppress("UNCHECKED_CAST")
                val citiesData = call.argument<List<Map<String, Any>>>("cities")
                if (citiesData != null) {
                    updateCities(citiesData)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Cities data required", null)
                }
            }
            else -> result.notImplemented()
        }
    }
    
    private fun updateCities(citiesData: List<Map<String, Any>>) {
        // 清除现有标记
        markers.forEach { it.remove() }
        markers.clear()
        cities.clear()
        
        // 解析城市数据
        for (cityData in citiesData) {
            val id = (cityData["id"] as? Number)?.toInt() ?: continue
            val name = cityData["name"] as? String ?: continue
            val lat = (cityData["latitude"] as? Number)?.toDouble() ?: continue
            val lng = (cityData["longitude"] as? Number)?.toDouble() ?: continue
            val country = cityData["country"] as? String ?: ""
            val score = (cityData["score"] as? Number)?.toDouble() ?: 0.0
            
            val city = CityMarkerData(id, name, lat, lng, country, score)
            cities.add(city)
            
            // 创建标记
            val markerColor = getMarkerColor(score)
            val markerOptions = MarkerOptions()
                .position(LatLng(lat, lng))
                .title(name)
                .snippet(country)
                .icon(BitmapDescriptorFactory.defaultMarker(markerColor))
            
            aMap?.addMarker(markerOptions)?.let { marker ->
                markers.add(marker)
            }
        }
        
        android.util.Log.d("AmapGlobal", "Added ${markers.size} city markers to AMap")
    }
    
    private fun getMarkerColor(score: Double): Float {
        return when {
            score >= 4.0 -> BitmapDescriptorFactory.HUE_GREEN
            score >= 3.0 -> BitmapDescriptorFactory.HUE_VIOLET
            else -> BitmapDescriptorFactory.HUE_RED
        }
    }
    
    private fun onMarkerClicked(marker: Marker) {
        // 查找对应的城市
        val index = markers.indexOf(marker)
        if (index >= 0 && index < cities.size) {
            val city = cities[index]
            methodChannel.invokeMethod("onCityTapped", mapOf(
                "id" to city.id,
                "name" to city.name,
                "latitude" to city.latitude,
                "longitude" to city.longitude,
                "country" to city.country,
                "score" to city.score
            ))
        }
    }
    
    // 生命周期方法
    fun onResume() {
        mapView.onResume()
    }
    
    fun onPause() {
        mapView.onPause()
    }
    
    fun onSaveInstanceState(outState: android.os.Bundle) {
        mapView.onSaveInstanceState(outState)
    }
}
